### PR TITLE
support rbd-nbd

### LINKED ISSUE #26: [RFE] separate node attach/detach/get device behind interface
By default krbd would be utilized, but it would be great if we could quickly and easily provide NBD and TCMU-based drivers if the node is configured to utilize something besides krbd.

### PR DESCRIPTION
**Pull request by rootfs**
fix #26

### CODE REVIEW

**Comment by rootfs** [comment]
@gman0 can you take a look? thanks

File: pkg/rbd/rbd_attach.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+// Check if rbd-nbd tools are installed.
+func checkRbdNbdTools() bool {
+	_, err := execCommand("modprobe", []string{"nbd"})
+	if err != nil {
+		glog.V(3).Infof("rbd-nbd: nbd modprobe failed with error %v", err)
+		return false
+	}
+	if _, err := execCommand("rbd-nbd", []string{"--version"}); err != nil {
+		glog.V(3).Infof("rbd-nbd: running rbd-nbd -h failed with error %v", err)
```
Comment (dillaman) [inline pkg/rbd/rbd_attach.go]: Nit: it's actually running ```--version```

File: pkg/rbd/rbd_attach.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+			return "", err
+		}
+
+		glog.V(3).Infof("rbd: map mon %s", volOptions.Monitors)
+		key, err := getRBDKey(userId, credentials)
+		if err != nil {
+			return "", err
+		}
+		output, err = execCommand(cmdName, []string{
+			"map", imagePath, "--id", userId, "-m", volOptions.Monitors, "--key=" + key})
```
Comment (dillaman) [inline pkg/rbd/rbd_attach.go]: Nit: I know this is an existing issue, but it doesn't seem safe to be passing the key in clear text. You can use the ```--keyfile``` optional to pass a path to a protected file which contains the key instead.

File: pkg/rbd/rbd_attach.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+			if err != nil {
+				return false, fmt.Errorf("fail to check rbd image status with: (%v), rbd output: (%s)", err, rbdOutput)
+			}
+			return !used, nil
+		})
+		// return error if rbd image has not become available for the specified timeout
+		if err == wait.ErrWaitTimeout {
+			return "", fmt.Errorf("rbd image %s is still being used", imagePath)
+		}
+		// return error if any other errors were encountered during wating for the image to becme avialble
```
Comment (gman0) [inline pkg/rbd/rbd_attach.go]: Typos in comment
> bcme

> avialble

File: pkg/rbd/rbd_attach.go [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+			if devicePath, found := getRbdDevFromImageAndPool(pool, image); found {
+				return devicePath, true
+			}
+		}
+	}
+	return "", false
+}
+
+// Check if rbd-nbd tools are installed.
+func checkRbdNbdTools() bool {
```
Comment (gman0) [inline pkg/rbd/rbd_attach.go]: Perhaps let the CO operator override the decision of the tool to be used (similar to what csi-cephfs is doing with `mounter`)?

File: pkg/rbd/rbd_attach.go (hunk 2 above)
Comment (rootfs) [reply pkg/rbd/rbd_attach.go]: yes, this is a long standing issue, especially in this case, rbdplugin has to pass /proc/pid/cmdline and get a chance to view the secret.

a potential fix is to let `rbd-nbd` print out the device path, so rbdplugin doesn't have to probe each nbd device through /proc to figure out which device maps to the image. I'll open a ceph issue for this feature.

**Comment by gman0** [comment]
Cool, please also update the rbd docs regarding both the storageclass param and the env variable

**Comment by rootfs** [comment]
added

**Pull request merged**